lib/adapters/apply-edit-adapter.ts

```diff
@@ -30,36 +42,35 @@ export default class ApplyEditAdapter {
   }
 
   public static async onApplyEdit(params: ApplyWorkspaceEditParams): Promise<ApplyWorkspaceEditResponse> {
-    let changes = params.edit.changes || {}
-
-    if (params.edit.documentChanges) {
-      changes = {}
-      params.edit.documentChanges.forEach((change) => {
-        if (change && "textDocument" in change && change.textDocument) {
-          changes[change.textDocument.uri] = change.edits
-        }
-      })
-    }
+    return ApplyEditAdapter.apply(params.edit)
+  }
 
-    const uris = Object.keys(changes)
+  public static async apply(workspaceEdit: WorkspaceEdit): Promise<ApplyWorkspaceEditResponse> {
+    ApplyEditAdapter.normalize(workspaceEdit)
 
     // Keep checkpoints from all successful buffer edits
     const checkpoints: Array<{ buffer: TextBuffer; checkpoint: number }> = []
 
-    const promises = uris.map(async (uri) => {
-      const path = Convert.uriToPath(uri)
-      const editor = (await atom.workspace.open(path, {
-        searchAllPanes: true,
-        // Open new editors in the background.
-        activatePane: false,
-        activateItem: false,
-      })) as TextEditor
-      const buffer = editor.getBuffer()
-      // Get an existing editor for the file, or open a new one if it doesn't exist.
-      const edits = Convert.convertLsTextEdits(changes[uri])
-      const checkpoint = ApplyEditAdapter.applyEdits(buffer, edits)
-      checkpoints.push({ buffer, checkpoint })
-    })
+    const promises = (workspaceEdit.documentChanges || []).map(
+      async (edit): Promise<void> => {
+        if (!TextDocumentEdit.is(edit)) {
+          return ApplyEditAdapter.handleResourceOperation(edit).catch((err) => {
+            throw Error(`Error during ${edit.kind} resource operation: ${err.message}`)
+          })
+        }
+        const path = Convert.uriToPath(edit.textDocument.uri)
+        const editor = (await atom.workspace.open(path, {
+          searchAllPanes: true,
+          // Open new editors in the background.
+          activatePane: false,
+          activateItem: false,
+        })) as TextEditor
```

Comment:
Any chance that this will return anything else?

Reply:
Not really I guess, but here's the interface
```typescript
open(uri: string, options?: WorkspaceOpenOptions): Promise<object>;
```

Reply:
Thank you. I might update the `@types/atom` later to reflect the edge cases.